### ISSUE TITLE
incoming Http2 to Http conversion behavior

### ISSUE THREAD

**Issue by matthpl** [netty/netty#4073]
I am considering building a proxy server using Netty with Spdy3.1 and Http2. I have some questions about Http2 to Http conversion.

For [InboundHttp2ToHttpAdapter](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/InboundHttp2ToHttpAdapter.java), it aggregates HEADERS frame and DATA frames to produce a FullHttpMessage. I am wondering the reason of doing this. Is it lack of use case or is it easier to understand? I also see this in Spdy to Http decoder.

For my proxy use case, when a client sends a proxy request using Http2,  the proxy would like to convert incoming HEADERS frame into a HttpMessage and send it upstream, then the proxy can use a HttpClient to send it to origin asap; if there are DATA frames follow the HEADERS frame, the proxy would convert DATA frames into HttpContent(chunks) and forward to origin. Building a FullHttpMessage may slow down the performance and I also worry about big POST requests.

**Comment by Scottmitch** [comment]
> For InboundHttp2ToHttpAdapter, it aggregates HEADERS frame and DATA frames to produce a FullHttpMessage. I am wondering the reason of doing this. Is it lack of use case or is it easier to understand?

The initial motivation behind this was "simplicity" w.r.t `Full*` objects defining the stream id and lifetime in 1 object (request payload was small / empty).  Since this was originally contributed it has not received much work because of direct usage of the HTTP/2 interfaces.  I would like to update the code to work with non `Full*` objects but I don't think I will have cycles for this any time soon.  We love PRs :)

**Comment by normanmaurer** [comment]
Yep that would be nice!

**Comment by matthpl** [comment]
It is very difficult to associate a HttpMessage and its HttpContent chunks, once they are converted from Http2 frames to http1.1 chunks.

Correct me if I am wrong. From my understanding, streams are FIFO and each stream is handled by separate thread.  No two streams would share the same handler within the same thread at the same time. Therefore, if we have a handler to convert a Http2 frames(for example stream Id 5) into Http1.1 request with chunks and pass them one by one to the next handler. The next handler will continue to serve the same Http1.1 request(was Http2 stream Id 5) until the last chunk. Once it receives last chunk, it will receive next Http1.1 request( coming from Http2 stream Id 7).

**Comment by Scottmitch** [comment]
> streams are FIFO

Not sure I understand the context here.  FIFO relative to what?

> each stream is handled by separate thread. No two streams would share the same handler within the same thread at the same time. 

The http2 codec currently assuming all operations are on the IO thread.

> if we have a handler to convert a Http2 frames(for example stream Id 5) into Http1.1 request 

Just to clarify the current code base
HTTP/2 -> HTTP/1.x is handled by [InboundHttp2ToHttpAdapter](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/InboundHttp2ToHttpAdapter.java) (already supports non `Full*` HTTP objects).
HTTP/1.x -> HTTP/2 is handled by [HttpToHttp2ConnectionHandler](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/HttpToHttp2ConnectionHandler.java) (what we need to consider here).

Just some high level thoughts related to this issue...We already have a header for which to communicate stream id in a [HTTP/1.x header](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/HttpUtil.java#L108).  So we could potentially extract this (or generate the next id if not present) and use this until a [LastHttpContent](https://github.com/netty/netty/blob/master/codec-http/src/main/java/io/netty/handler/codec/http/LastHttpContent.java).

**Comment by mosesn** [comment]
@Scottmitch @matthpl @normanmaurer I might be interested in picking this up, but I'm not sure if it's feasible for the general case.  Fundamentally the problem is dealing with an `HttpContent` that's not an `HttpRequest` or a `LastHttpContent` because there isn't anywhere to shove the stream id.

A few possible strategies I've considered:

1.  Provide an `HttpContent` proxy that has an extra method that exposes a stream id.  The downside here is if for some reason there's a handler that inserts its own messages, or aggregates or chunks and doesn't know about stream ids.  It will actually work with HttpObjectAggregator because the FullHttpRequest will still have the right headers, but it might fail with something else.

2.  Shove the stream id into the ChannelHandlerContext AttributeMap, and then read out of it when converting to http/2.  The downside here is that it won't work if we ever buffer writes.

What do you think?  They're both pretty hacky, do you think it's even worth doing while we wait for https://github.com/netty/netty/issues/4913?

**Comment by Scottmitch** [comment]
@mosesn - Sorry for the delay. I would recommend focusing efforts on the child channel API / implementation.

**Comment by mosesn** [comment]
@Scottmitch yeah, I basically had to reimplement child channels to get something working on my end.  seems reasonable.

**Comment by hyperxpro** [comment]
I am working on this for my Load Balancer and I also need this feature. I will do PR soon. Just set `boolean` to `true` in `InboundHttp2ToHttpAdapter` and it'll convert Http/2 frames to HttpObjects (HttpMessage and HttpContent) instead of FullHttpMessage.

**Comment by hyperxpro** [comment]
Finally, after 5 years, this feature is coming. :D

@normanmaurer  @Scottmitch PTAL #10636
